Options for filename and truncation of imports

chooban opened the pull request:
Thank you very much for the package, we've found it very helpful.

We'd like to be able to generate two guard files, one with debug output and one without, so that we can switch out one for the other when in development mode vs the production build. The swapping is taken care of by our build process, but the hardcoded filenames and full import paths were causing issues.

The option to change the filename means we can generated `.debugguard.ts` and `.guard.ts` files. The imports were an issue since it'll change with either a developer's, or CI's, workspace location.

I hope the changes below are suitable, happy to make changes.

rhys-vdw commented:
Hey @chooban, sorry mate just saw this one. Thanks for the PR.

An option to change the `.guard.ts` in output files is good, and I think this is the right way to approach the problem (run ts-auto-guard twice).

For this to get merged I'll need:
1. Inclusion of flags in README file
2. Test coverage for new features.
3. Tell me more about the "truncate imports" option and its purpose?

chooban commented:
Merged up to date, updated the readme and added a couple of tests.

The truncation one is probably poorly named. The scenario I'm thinking of is having two files generated for the guards, one debug and one not. We would use the debug one at dev time and the non-debug one in production.

The issue with that is that the debug statements for custom interfaces contain import paths that give a full path to the file containing the interface. So if Developer A generates the guards and Developer B does the same without changing the interface, the debug output will still change as the path will be difference. 

The option is a naive fix which removes the current working directory from any import paths it detects, hopefully leaving it in a state that means we won't have unneccessary file churn. In particular, we have a CI job that regenerates the guards at PR time and ensures there are no forgotten changes, if the import path is included without this then it'll fail every time.

It's a niche issue, I admit! :D

rhys-vdw commented:
Great stuff @chooban. Thank you for the comprehensive test coverage.

> The option is a naive fix which removes the current working directory from any import paths it detects, hopefully leaving it in a state that means we won't have unneccessary file churn.

So are you say that at present you get the full path like `/Users/username/Projects/my-cool-project/src/types/MyType` in the require (rather than `src/types/MyType`)? If so then we should _always_ be using paths relative to the project root, and this doesn't need to be an option.

Don't get the impression that every behaviour of this tool is intentional hahah. It's very WIP. I'm going to promote you to collaborator now, but I'm not going to merge it because, if I understand correctly, we don't need a `--truncate-imports` option, as that should _always_ happen.

Let me know your thoughts. If you cbf or don't trust the reliability of the feature then pls open an issue like "Consider making --truncate-imports default and non-optional", and then go ahead and merge.

chooban commented:
I've made truncation the default and pushed it up. Presuming every passes I'll merge it in. It's a naive approach at the moment, but it'll be one of those things that I hope works for 99% of scenarios.
